consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/FutureMessageBuffer.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft.statemachine;
+
+import tech.pegasys.pantheon.ethereum.p2p.api.Message;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Buffer which holds future IBFT messages.
+ *
+ * <p>This buffer only allows messages to be added which have a chain height greater than current
+ * height and up to chain futureMessagesMaxDistance from the current chain height.
+ *
+ * <p>When the total number of messages is greater futureMessagesLimit then messages are evicted.
+ *
+ * <p>If there is more than one height in the buffer then all messages for the highest chain height
+ * are removed. Otherwise if there is only one height the oldest inserted message is removed.
+ */
+public class FutureMessageBuffer {
+  private final NavigableMap<Long, List<Message>> buffer = new TreeMap<>();
+  private final long futureMessagesMaxDistance;
+  private final long futureMessagesLimit;
+
+  public FutureMessageBuffer(final long futureMessagesMaxDistance, final long futureMessagesLimit) {
+    this.futureMessagesMaxDistance = futureMessagesMaxDistance;
+    this.futureMessagesLimit = futureMessagesLimit;
+  }
+
+  public void addMessage(final long chainHeight, final long msgChainHeight, final Message rawMsg) {
```

Comment:
don't LOVE msgChainHeight coming in on this API .... rather this get reset in the Poll somehow.

Reply:
done.changed so it is using been passed in as a arg.